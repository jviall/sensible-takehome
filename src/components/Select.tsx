import { useId } from "react";
import { ILocation } from "../utils/types";

interface Props {
  value: string;
  onChange: React.ChangeEventHandler<HTMLSelectElement>;
  options: ILocation[];
  label: string;
}

export default function Select({ value, onChange, options, label }: Props) {
  const id = useId();

  return (
    <div className="flex flex-col">
      <label htmlFor={id} className="mr-2">
        {label}
      </label>
      <select
        id={id}
        onChange={onChange}
        value={value}
        className="rounded-md p-1"
      >
        {options.map((opt) => (
          <option value={JSON.stringify(opt)} key={opt.name}>
            {opt.name}
          </option>
        ))}
      </select>
    </div>
  );
}
