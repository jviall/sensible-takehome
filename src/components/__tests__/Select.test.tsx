import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useState } from "react";
import Select from "../Select";
import { MOCK_LOCATIONS } from "./data";

const Fixture = (props: Parameters<typeof Select>[0]) => {
  const [value, setValue] = useState(props.value);
  return (
    <Select
      {...props}
      value={value}
      onChange={(e) => {
        setValue(e.target.value);
        props.onChange(e);
      }}
    />
  );
};

test("Select matches base snapshot", () => {
  const { asFragment } = render(
    <Fixture
      onChange={vi.fn()}
      value={JSON.stringify(MOCK_LOCATIONS[0])}
      label="Search"
      options={MOCK_LOCATIONS}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});

test("Select accepts a default value that isn't the first option", () => {
  const defaultValue = JSON.stringify(MOCK_LOCATIONS[1]);
  const { getByRole } = render(
    <Fixture
      onChange={vi.fn()}
      value={defaultValue}
      label="Search"
      options={MOCK_LOCATIONS}
    />
  );
  const option = getByRole("option", {
    name: MOCK_LOCATIONS[1].name,
  }) as HTMLOptionElement;
  expect(option.selected).toBe(true);
});

test("Select value responds to users choice", async () => {
  const user = userEvent.setup();
  const onChange = vi.fn();
  const defaultValue = JSON.stringify(MOCK_LOCATIONS[0]);
  const expectedValue = JSON.stringify(MOCK_LOCATIONS[1]);
  const { getByRole } = render(
    <Fixture
      onChange={onChange}
      value={defaultValue}
      label="Search"
      options={MOCK_LOCATIONS}
    />
  );

  const input = getByRole("combobox");
  const option = getByRole("option", {
    name: MOCK_LOCATIONS[1].name,
  }) as HTMLOptionElement;
  await user.selectOptions(input, expectedValue);

  expect(option.selected).toBe(true);
  expect(onChange).toHaveBeenCalledOnce();
});
