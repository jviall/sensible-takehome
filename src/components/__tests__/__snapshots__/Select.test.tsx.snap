// Vitest Snapshot v1

exports[`Select matches base snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
  >
    <label
      class="mr-2"
      for=":r0:"
    >
      Search
    </label>
    <select
      class="rounded-md p-1 bg-white dark:bg-neutral-700"
      id=":r0:"
    >
      <option
        value="{\\"name\\":\\"Pizza\\",\\"center\\":{\\"lat\\":0,\\"lng\\":0}}"
      >
        Pizza
      </option>
      <option
        value="{\\"name\\":\\"Tacos\\",\\"center\\":{\\"lat\\":1,\\"lng\\":1}}"
      >
        Tacos
      </option>
      <option
        value="{\\"name\\":\\"Pasta\\",\\"center\\":{\\"lat\\":2,\\"lng\\":2}}"
      >
        Pasta
      </option>
    </select>
  </div>
</DocumentFragment>
`;
